WIP feat(#2478): add work side menu component

bdfranck opened the pull request:
This PR adds the work side menu component.

https://github.com/user-attachments/assets/7406430a-e93c-47dc-bd86-86f00a3470a8

### To do
- [ ] Decide component name
- [ ] Decide property names
- [x] Create react wrapper
- [x] Create angular wrapper
- [ ] Ensure that angular and react align with v4 changes
- [x] Add initial component tokens
- [x] Add ARIA labels
- [ ] Write tests
- [ ] Write docs
- [x] Rename commit message to include issue number
- [ ] Test different variations
- [x] Cross-browser testing

### How to test

Add one of the following menus to your playground:

#### Empty menu
```svelte
<goa-work-side-menu
  url="#"
  title="Service name"
  username="Name"
  useremail="test@example.com"
>
</goa-work-side-menu>
``` 

#### Menu with primary, secondary, and account items
```svelte
<goa-work-side-menu
  url="#"
  title="Service name"
  username="Name"
  useremail="test@example.com"
>
  <div slot="primary">
    <goa-work-side-menu-item
      icon="triangle"
      label="Primary menu item"
      url="#primary"
    />
  </div>
  <div slot="secondary">
    <goa-work-side-menu-item
      icon="triangle"
      label="Secondary menu item"
      url="#secondary"
    />
  </div>
  <div slot="account">
    <goa-work-side-menu-item
      icon="triangle"
      label="Account menu item"
      url="#secondary"
    />
  </div>
</goa-work-side-menu>
``` 

#### Menu with primary , secondary, account items, and badges
```svelte
<goa-work-side-menu
  url="#"
  title="Service name"
  username="Name"
  useremail="test@example.com"
>
  <div slot="primary">
    <goa-work-side-menu-item
      icon="triangle"
      label="Primary menu item"
      url="#primary"
      badge="New
    />
  </div>
  <div slot="secondary">
    <goa-work-side-menu-item
      icon="triangle"
      label="Secondary menu item"
      url="#secondary"
      type="emergency"
      badge="12"
    />
  </div>
  <div slot="account">
    <goa-work-side-menu-item
      icon="triangle"
      label="Account menu item"
      url="#secondary"
    />
  </div>
</goa-work-side-menu>
``` 

#### A full example with a sub menu

```svelte
<goa-work-side-menu
  url="#client"
  title="Income and Employment Support (IES)"
  username="Edna Mode"
  useremail="edna.mode@example.com"
>
  <div slot="primary">
    <goa-work-side-menu-item
      icon="search"
      label="Search"
      badge="21"
      url="#search"
    >
      <goa-work-side-menu-item
        url="#sub1"
        variation="submenu"
        label="Sub menu item"
      >
      </goa-work-side-menu-item>

      <goa-work-side-menu-item
        url="#sub2"
        variation="submenu"
        label="Sub menu item"
      />

      <goa-work-side-menu-item
        url="#sub3"
        variation="submenu"
        label="Sub menu item"
      />
    </goa-work-side-menu-item>

    <goa-work-side-menu-item
      icon="list"
      label="Clients"
      type="success"
      badge="New"
      current="true"
      url="#clients"
    />

    <goa-work-side-menu-item
      icon="calendar"
      label="Schedule"
      type="emergency"
      badge="Urgent"
      url="#schedule"
    />
  </div>

  <div slot="secondary">
    <goa-work-side-menu-item
      icon="notifications"
      label="Notifications"
      type="success"
      badge="1"
      badgeLabelType="alwaysvisible"
      url="#notifications"
    />

    <goa-work-side-menu-item icon="help-circle" label="Support" url="#support" />

    <goa-work-side-menu-item icon="settings" label="Settings" url="#settings" />
  </div>

  <div slot="account">
    <goa-work-side-menu-item
      icon="person"
      label="Account management"
      url="#account"
    />

    <goa-work-side-menu-item icon="log-out" label="Log out" url="#logout" />
  </div>
</goa-work-side-menu>
```

bdfranck commented on libs/web-components/src/components/work-side-menu/WorkSideMenu.svelte:
```diff
@@ -0,0 +1,691 @@
+  // Public
+  // ******
+
+  export let title: string;
```
I've changed the property to `heading` to match the App Header Menu.

chrisolsen commented on libs/react-components/src/lib/work-side-menu-item/work-side-menu-item.spec.tsx:
```diff
@@ -0,0 +1,19 @@
+import { render } from "@testing-library/react";
```
@bdfranck re: alignment => Yes, things would shift/padding would be reduced by a couple of pixels.

bdfranck commented:
@chrisolsen I addressed all your comments and pushed an amended commit. I've also added browser tests for testing events in the menu and menu item. Does the location of the Angular `experimental` folder make sense? Any further questions or concerns? Thanks!

chrisolsen commented on libs/web-components/src/components/work-side-menu/WorkSideMenuWrapper.test.svelte:
```diff
@@ -0,0 +1,33 @@
+<svelte:options customElement="test-work-side-menu-wrapper" />
```
We don't need the *wrapper* tests any more now that we have the browser tests.

chrisolsen commented on libs/web-components/src/components/work-side-menu/WorkSideMenuItem.svelte:
```diff
@@ -0,0 +1,334 @@
+      right: var(--goa-space-3xs);
+      text-align: center;
+      line-height: 1.2;
+      font-size: var(--goa-font-size-1, 12px);
```
Is there a reason there is a fallback value for this one? If so, it should probably fall back to a `rem` value.

chrisolsen commented on libs/web-components/src/components/work-side-menu/WorkSideMenuItem.svelte:
```diff
@@ -0,0 +1,334 @@
+  /* Badge */
+  .badge {
+    color: var(--goa-color-text-light);
+    height: 20px;
```
We need to change a lot of these px values to rem.

bdfranck replied:
I've removed every `px` value that has a relevant token value.

chrisolsen commented on libs/react-components/specs/work-side-menu-item.browser.spec.tsx:
```diff
@@ -0,0 +1,63 @@
+import { render } from "vitest-browser-react";
```
I am unsure we want to use browser tests for testing *child* components.

chrisolsen commented on libs/web-components/src/components/work-side-menu/WorkSideMenu.svelte:
```diff
@@ -0,0 +1,748 @@
+  export let heading: string;
+  export let url: string;
+  export let username: string;
+  export let useremail: string;
```
As per some other designs I have seen we need something less specific here as in the other examples I believe I saw the users `role` being used.

chrisolsen commented on libs/web-components/src/components/work-side-menu/WorkSideMenu.svelte:
```diff
@@ -0,0 +1,748 @@
+
+  export let heading: string;
+  export let url: string;
+  export let username: string;
```
For the properties, we should probably start shifting over to using the kebab convention. This will require some changes to the `customElement` prop at the top of the page. Check out the Link.svelte file for an example on how to do this.

bdfranck replied:
I've added kebab case for both user attributes.

twjeffery reviewed (changes requested):
@bdfranck This component is looking great. A few things for us to consider:

**1. Back Button / Secondary Navigation:** When the side menu is used as a secondary nav for something like a _detail view_, we add a _back_ menu button that's a bit spaced out away from the rest of the main menu items. Should we add this as a property for teams to be able to do this easily? Either as a type of menu item or as a prop on the menu itself (such as `has-back-button` or `is-secondary`)

<img width="182" height="355" alt="image" src="https://github.com/user-attachments/assets/19b345d2-9f3b-4243-ad69-483ebbf12472" />


**2. Submenu Expand/Collapse Behavior**
 - Currently submenus only show when parent item is active. (auto-revealed on navigation). It makes it a bit hard to know where the sub items are until you reveal that item. Should we add a chevron to toggle the menu open and closed? 
    - Perhaps something like this, but it's not fully resolved
    
      <img width="314" height="399" alt="image" src="https://github.com/user-attachments/assets/f8f8b23c-deec-4b46-a38a-90ba5b832f94" />

**3. Menu State Persistence During Navigation**
 - Should collapsed/expanded state persist when user navigates between pages? (Perhaps with the exception if there are sub items). Currently the menu opens whenever you navigate, which feels a little cumbersome if you've just collapsed the menu.

**4. The keyboard focus** does not enter into the Account/profile menu when it opens. And pressing `Escape` should close it

---
Here is a jam.dev of what I checked and adjusted while testing:
https://jam.dev/c/c212d2f9-e99e-4d1c-a72f-59fa5721e74e

twjeffery commented on libs/web-components/src/components/work-side-menu/WorkSideMenu.svelte:
```diff
@@ -0,0 +1,800 @@
+      var(--goa-border-width-s) solid var(--goa-color-greyscale-200)
+    );
+    background-color: transparent;
+    padding: var(--goa-space-xs) var(--goa-space-2xs) var(--goa-space-xs)
```
Adjust to `padding: var(--goa-space-xs);` to match closed spacing to avoid shift on expand and collapse, and icons line up in both states.

<img width="314" height="286" alt="image" src="https://github.com/user-attachments/assets/58ad6d7c-c820-4f78-893f-a5f98db27aa5" />

<img width="314" height="286" alt="image" src="https://github.com/user-attachments/assets/3fef7ef6-60d3-4e29-9b55-7baad0af029e" />

twjeffery commented:
@bdfranck

**1. Back button** 
Sounds good

**2. Submenu items**
Sounds good

**3. Menu state persistence**
For sure, it might be a playground issue that wouldn't be an issue for teams. I am using multiple pages in my playground.

**4. Keyboard navigation for the account menu**
For sure. Maybe this should be pulled out to a separate issue for the popover?

twjeffery reviewed (changes requested):
Great work @bdfranck, I pulled in the changes and noticed a few more small things, added as comments

chrisolsen commented on libs/angular-components/src/index.ts:
```diff
@@ -1,3 +1,4 @@
 export * from "./lib/angular-components.module";
 export * from "./lib/components";
 export * from "@abgov/ui-components-common";
+export * from "./experimental";
```
Having this line here will bundle the new experimental component with all the other components. In the `dist/libs/angular-components` folder is there an `experimental/index.js` file? If so this line should be able to be removed.

twjeffery reviewed (approved):
@bdfranck Just one small potential update, looks good

vanessatran-ddi commented on libs/react-components/src/experimental/work-side-menu/work-side-menu.tsx:
```diff
@@ -0,0 +1,78 @@
+  primaryContent?: ReactNode;
+  secondaryContent?: ReactNode;
+  accountContent?: ReactNode;
+  open?: boolean;
```
If we want to pass `open` as boolean, then under WorkSideMenu.svelte we need to assign it as type boolean, and by default `let open = false;` (you can refer Drawer.svelte), otherwise you can assign `open` as string, and follow the way `Modal.svelte` and `modal.tsx` doing, otherwise, the `open` is always `true` here. I need to add some code to make it works locally 
<img width="1035" height="410" alt="image" src="https://github.com/user-attachments/assets/0d2ca63b-39ab-448d-be0e-8b77452dd640" />

vanessatran-ddi commented:
> @vanessatran-ddi I've amended my commit with the following changes:
> 
> * Added `experimental` to the Angular imports
> * Added `aria-expaned` to the account menu
> * Improved keyboard navigation
>   
>   * Works on entire menu
>   * Loops when reaching the bottom or top
>   * Opens the account menu when navigating to it
>   * Closes the account menu when navigating from it
> * Fixed the border between the header and main menu when scrolling (an issue that @Spark450 noted)
> 
> ![keyboard-nav](https://private-user-images.githubusercontent.com/1479091/503270418-5a2ece2b-2d6c-4ebf-82c4-f75c2d3273ee.gif?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NjA5OTg0NjcsIm5iZiI6MTc2MDk5ODE2NywicGF0aCI6Ii8xNDc5MDkxLzUwMzI3MDQxOC01YTJlY2UyYi0yZDZjLTRlYmYtODJjNC1mNzVjMmQzMjczZWUuZ2lmP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MTAyMCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTEwMjBUMjIwOTI3WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9OTFiOGNmMjRkYjg4NGEwMWM0MDA3ZTQxMjE0ZDg4N2FkMzcwOGFmMzU5MTg1MThkNzI4Y2NjNDhhZGNiYjlkYSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ._h200WseDoA41pZZaxuXclIKaUgVFnnz2isrQN7igVk) [ ![keyboard-nav](https://private-user-images.githubusercontent.com/1479091/503270418-5a2ece2b-2d6c-4ebf-82c4-f75c2d3273ee.gif?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NjA5OTg0NjcsIm5iZiI6MTc2MDk5ODE2NywicGF0aCI6Ii8xNDc5MDkxLzUwMzI3MDQxOC01YTJlY2UyYi0yZDZjLTRlYmYtODJjNC1mNzVjMmQzMjczZWUuZ2lmP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MTAyMCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTEwMjBUMjIwOTI3WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9OTFiOGNmMjRkYjg4NGEwMWM0MDA3ZTQxMjE0ZDg4N2FkMzcwOGFmMzU5MTg1MThkNzI4Y2NjNDhhZGNiYjlkYSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ._h200WseDoA41pZZaxuXclIKaUgVFnnz2isrQN7igVk) ](https://private-user-images.githubusercontent.com/1479091/503270418-5a2ece2b-2d6c-4ebf-82c4-f75c2d3273ee.gif?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NjA5OTg0NjcsIm5iZiI6MTc2MDk5ODE2NywicGF0aCI6Ii8xNDc5MDkxLzUwMzI3MDQxOC01YTJlY2UyYi0yZDZjLTRlYmYtODJjNC1mNzVjMmQzMjczZWUuZ2lmP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MTAyMCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTEwMjBUMjIwOTI3WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9OTFiOGNmMjRkYjg4NGEwMWM0MDA3ZTQxMjE0ZDg4N2FkMzcwOGFmMzU5MTg1MThkNzI4Y2NjNDhhZGNiYjlkYSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ._h200WseDoA41pZZaxuXclIKaUgVFnnz2isrQN7igVk) [ ](https://private-user-images.githubusercontent.com/1479091/503270418-5a2ece2b-2d6c-4ebf-82c4-f75c2d3273ee.gif?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NjA5OTg0NjcsIm5iZiI6MTc2MDk5ODE2NywicGF0aCI6Ii8xNDc5MDkxLzUwMzI3MDQxOC01YTJlY2UyYi0yZDZjLTRlYmYtODJjNC1mNzVjMmQzMjczZWUuZ2lmP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MTAyMCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTEwMjBUMjIwOTI3WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9OTFiOGNmMjRkYjg4NGEwMWM0MDA3ZTQxMjE0ZDg4N2FkMzcwOGFmMzU5MTg1MThkNzI4Y2NjNDhhZGNiYjlkYSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ._h200WseDoA41pZZaxuXclIKaUgVFnnz2isrQN7igVk)
> 
> <img alt="image" width="330" height="201" src="https://private-user-images.githubusercontent.com/1479091/503270484-f705304d-4623-450b-b267-1b3712de645f.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NjA5OTg0NjcsIm5iZiI6MTc2MDk5ODE2NywicGF0aCI6Ii8xNDc5MDkxLzUwMzI3MDQ4NC1mNzA1MzA0ZC00NjIzLTQ1MGItYjI2Ny0xYjM3MTJkZTY0NWYucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MTAyMCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTEwMjBUMjIwOTI3WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ZjQ0MDFjNzU2ZjU0ZjM0ZGIwM2E5OTc3M2U3MjhlNTQzY2E3MzIxNGI2YjA2ZGYzNzliMzZmMDI2MjgxOWFiMyZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.otjHxGJArFw4vZDehwmAfXdwPawJZSpgKZu2gzgIg0E">

Hi @bdfranck , 
I have this comment to follow up vs you: 

Current behavior: Arrow Down from "Edna Mode" automatically expands the submenu and focuses on "Account management" → "Log out" before continuing to "Collapse Menu".

Expected behavior: Arrow Down from "Edna Mode" should go directly to "Collapse Menu" without auto-expanding the submenu. The user should need to press Enter/Space on "Edna Mode" to expand and access the submenu items.

chrisolsen commented on libs/react-components/specs/work-side-menu.browser.spec.tsx:
```diff
@@ -0,0 +1,189 @@
+      const result = render(<Component />);
+
+      await vi.waitFor(() => {
+        expect(result.getByTestId("work-side-menu")).toBeTruthy();
```
As per the previous comment this can be moved out of this waitFor block

bdfranck commented:
Here's a summary of some decisions I made for this component:
- I chatted with Tom about the issue where the tooltip and popover menu don't appear aligned if the menu is placed in an unconventional position. The only valid use of the component is if it's placed in the top left. We couldn't come up with a valid scenario of using it in an unconventional position. We decided to leave the issue and return to it later if a scenario arises.
- I removed the submenu items and created a separate ticket for a [side menu group](https://github.com/GovAlta/ui-components/issues/3137). I decided this because the existing submenu had accessibility and interaction issues. It was attempting to do the job of both a single menu item and a menu group at the same time. A dedicated menu group component felt like a better approach.

chrisolsen commented on libs/react-components/specs/work-side-menu.browser.spec.tsx:
```diff
@@ -0,0 +1,189 @@
+      // Wait for component to be fully rendered
+      await vi.waitFor(() => {
+        expect(result.getByTestId("work-side-menu")).toBeTruthy();
+      });
```
No need for the waitFor and expect check. This applies to all other similar code blocks, so I will stop commenting on these :)

bdfranck commented:
@chrisolsen I've amended my commit with the following changes:
- Changed all instances of `goax` to `goabx`
- Removed all the unneeded `waitFor`s in the browser test
- Rebased from `alpha`

vanessatran-ddi commented on libs/angular-components/src/experimental/work-side-menu/work-side-menu.ts:
```diff
@@ -0,0 +1,53 @@
+import {
```
QA:
I tested on angular with the following case:
1. I tab to focus on the workspace side menu
2. I press Arrow down
3. I expected to see my focusing menu item to be the first menu "Dashboard" but in fact it is "John Doe" and when I press Arrow down again, it comes to Collapse Menu. Arrow up and down only moving between John Doe & Collapse Menu

I added some console.log on my local machine, and I think  for angular, you will need to add a small delay (refer how other angular components doing such as drawer.ts), otherwise Svelte will `onMount` too quickly, that doesn't register all `goabx-work-side-menu-item` correctly to the `_focusItems`

<img width="851" height="339" alt="image" src="https://github.com/user-attachments/assets/e19c993f-4724-489d-b185-524dff424ae1" />


https://github.com/user-attachments/assets/5cbe9e37-aa2b-446b-9780-c692fe41fdd2

vanessatran-ddi replied:
Also for Angular, when I press Enter to open the User's popover:
<img width="951" height="1107" alt="image" src="https://github.com/user-attachments/assets/c00abbf6-cf74-4b9a-97f0-e1c6bf33f228" />

In React, if I arrow up and down, in some moments, I can focus on the menu items inside the user's popover. For Angular, I press Arrow up and down, no matter how many times I press, I never have a chance to focus on the menu item inside the user's popover. I think the problem is similar to `missing a delay` for angular to be able to register vs Svelte's onMount

vanessatran-ddi commented on libs/angular-components/src/experimental/work-side-menu/work-side-menu.ts:
```diff
@@ -0,0 +1,69 @@
+@Component({
+  standalone: true,
+  selector: "goabx-work-side-menu", // eslint-disable-line
+  imports: [NgTemplateOutlet],
```
I have this error: 
<img width="549" height="290" alt="image" src="https://github.com/user-attachments/assets/330aac4c-22d8-4416-a3b9-b691caccb4b4" />


I think you are missing CommonModule here.

vanessatran-ddi commented on libs/react-components/src/experimental/work-side-menu/work-side-menu.tsx:
```diff
@@ -0,0 +1,78 @@
+export interface GoabWorkSideMenuProps {
+  heading: string;
+  url: string;
+  userName: string;
```
`heading` in angular is optional, but `heading` under react is mandatory, same vs `url`, `username`
We should follow Angular for React, `url` or `userName` shouldn't be mandatory.

vanessatran-ddi replied:
Same here, we should make `userName` and `userSecondaryText` optional (Example our new design system website DS 2.0 will use this workspace menu and won't need `userName`, `userSecondaryText`.

vanessatran-ddi commented on libs/web-components/src/components/work-side-menu/WorkSideMenu.svelte:
```diff
@@ -0,0 +1,895 @@
+  }
+
+  // Account menu
+  function closeAccountMenu(e: Event) {
```
This function is being called multiple times
<img width="1047" height="327" alt="image" src="https://github.com/user-attachments/assets/17a3a971-0cd5-4241-b757-7403e0ee8680" />

However, I can see we have duplicated code under Escape: 
<img width="870" height="177" alt="image" src="https://github.com/user-attachments/assets/e27b3c6d-9ffd-4633-8c67-72018d5a5761" />


My recommendation is: if we have multiple places call this function to remove the `click` listener, there will be cases that we call and remove the listener that even not exists. We should add a check like what you did under `Escape` and replace the duplicated code vs closeAccountMenu

```
function closeAccountMenu(e?: Event) {
    if (_showAccountMenu) {
      _showAccountMenu = false;
      window.removeEventListener("click", closeAccountMenu);
    }
  }

```

vanessatran-ddi commented on libs/web-components/src/components/work-side-menu/WorkSideMenu.svelte:
```diff
@@ -0,0 +1,895 @@
+  let _mouseEnterTimeoutId: any;
+  let _mouseLeaveTimeoutId: any;
+
+  let _windowWidth = window.innerWidth;
```
This variable is not being used, we should remove.

vanessatran-ddi commented on libs/web-components/src/components/work-side-menu/WorkSideMenu.svelte:
```diff
@@ -0,0 +1,895 @@
+  }
+
+  function getShadowLinkEl(el: Element): HTMLAnchorElement | null | undefined {
+    if (el.tagName.includes("GOABX")) el = el.firstElementChild as Element;
```
if `el.firstElementChild` is null -> we will force Typescript to treat `null` as `Element`. 
Perhaps: 
```
function getShadowLinkEl(el: Element): HTMLAnchorElement | null | undefined {
    if (el.tagName.includes("GOABX")) {
      const firstChild = el.firstElementChild;
      if (!firstChild) {
        return null;  // Early return if no child exists
      }
      el = firstChild;
    }

    return el?.shadowRoot?.querySelector("a");
  }

```

vanessatran-ddi commented on libs/web-components/src/components/work-side-menu/WorkSideMenu.svelte:
```diff
@@ -0,0 +1,895 @@
+  }
+
+  function setFocusedIndexToElement(el: HTMLElement) {
+    _focusedIndex = _focusItems.indexOf(el) || -1;
```
This logic will be wrong if `indexOf(e)` returns the first `index` which is `0`
<img width="555" height="178" alt="image" src="https://github.com/user-attachments/assets/bca8b43a-5beb-4a95-a30b-170ff68e7b2d" />

vanessatran-ddi commented on libs/web-components/src/components/work-side-menu/WorkSideMenu.svelte:
```diff
@@ -0,0 +1,895 @@
+    }
+  }
+
+  function onArrow(e: KeyboardEvent, direction: "up" | "down") {
```
We can remove this `e`
<img width="742" height="176" alt="image" src="https://github.com/user-attachments/assets/781674b5-520f-4276-a94d-7b1588a363c7" />

vanessatran-ddi commented on libs/web-components/src/components/work-side-menu/WorkSideMenuItem.svelte:
```diff
@@ -0,0 +1,289 @@
+
+<script lang="ts">
+  import { onMount, onDestroy, tick } from "svelte";
+  import { dispatch, getSlottedChildren } from "../../common/utils";
```
We can remove `getSlottedChildren` since it is not being used.

vanessatran-ddi commented on libs/web-components/src/components/work-side-menu/WorkSideMenuItem.svelte:
```diff
@@ -0,0 +1,289 @@
+>
+  <a
+    class="menu-item"
+    class:current
```
Consider to use `aria-current="page"` when `current` is true. 
Source: https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Reference/Attributes/aria-current

vanessatran-ddi commented on libs/web-components/src/components/work-side-menu/WorkSideMenuItem.svelte:
```diff
@@ -0,0 +1,289 @@
+    on:click={handleClick}
+    tabindex="0"
+  >
+    <goa-icon size="small" theme={current ? "filled" : "outline"} type={icon} />
```
Same here, add `arialabel` so when the menu is collapsed, the screen reader is still able to announce the menu. I suggest to use `label` as the value.

vanessatran-ddi commented on libs/web-components/src/components/work-side-menu/WorkSideMenu.svelte:
```diff
@@ -0,0 +1,906 @@
+            on:click={handleToggleClick}
+            bind:this={_toggleButtonEl}
+          >
+            <goa-icon
```
Missing the `ariaLabel` for "Collapse Menu" or "Expand Menu" here, otherwise when we focus on Expand/Collapse button, we only hear that "image, menu, collapse button".

vanessatran-ddi commented on libs/web-components/src/components/work-side-menu/WorkSideMenu.svelte:
```diff
@@ -0,0 +1,907 @@
+      }
+    });
+    observer.observe(document.body, { childList: true, subtree: true });
+    return observer; // Return the created observer
```
we can remove the comment here.

vanessatran-ddi reviewed (commented):
Several comments left. If you fix them let me know and I am ready to approve.

tzuge commented:
:tada: This PR is included in version 1.40.0-dev.10 :tada:

The release is available on:
- [npm package (@dev dist-tag)](https://www.npmjs.com/package/@abgov/web-components/v/1.40.0-dev.10)
- [GitHub release](https://github.com/GovAlta/ui-components/releases/tag/web-components-v1.40.0-dev.10)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:

tzuge commented:
:tada: This PR is included in version 6.10.0-dev.4 :tada:

The release is available on:
- [npm package (@dev dist-tag)](https://www.npmjs.com/package/@abgov/react-components/v/6.10.0-dev.4)
- [GitHub release](https://github.com/GovAlta/ui-components/releases/tag/react-v6.10.0-dev.4)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:

chrisolsen commented:
:tada: This PR is included in version 1.10.0-next.1 :tada:

The release is available on:
- [npm package (@next dist-tag)](https://www.npmjs.com/package/@abgov/ui-components-common/v/1.10.0-next.1)
- [GitHub release](https://github.com/GovAlta/ui-components/releases/tag/common-v1.10.0-next.1)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:

chrisolsen commented:
:tada: This PR is included in version 4.10.0-next.1 :tada:

The release is available on:
- [npm package (@next dist-tag)](https://www.npmjs.com/package/@abgov/angular-components/v/4.10.0-next.1)
- [GitHub release](https://github.com/GovAlta/ui-components/releases/tag/angular-v4.10.0-next.1)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:

chrisolsen commented:
:tada: This PR is included in version 6.10.0-next.1 :tada:

The release is available on:
- [npm package (@next dist-tag)](https://www.npmjs.com/package/@abgov/react-components/v/6.10.0-next.1)
- [GitHub release](https://github.com/GovAlta/ui-components/releases/tag/react-v6.10.0-next.1)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:

chrisolsen commented:
:tada: This PR is included in version 1.40.0 :tada:

The release is available on:
- [npm package (@latest dist-tag)](https://www.npmjs.com/package/@abgov/web-components/v/1.40.0)
- [GitHub release](https://github.com/GovAlta/ui-components/releases/tag/web-components-v1.40.0)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:
